src/System.Private.Uri/src/System/UriExt.cs

```diff
@@ -499,29 +499,46 @@ internal unsafe bool InternalIsWellFormedOriginalString()
         public static string UnescapeDataString(string stringToUnescape)
         {
             if ((object)stringToUnescape == null)
+            {
                 throw new ArgumentNullException(nameof(stringToUnescape));
+            }
 
             if (stringToUnescape.Length == 0)
+            {
                 return string.Empty;
+            }
 
             unsafe
             {
                 fixed (char* pStr = stringToUnescape)
                 {
                     int position;
                     for (position = 0; position < stringToUnescape.Length; ++position)
+                    {
                         if (pStr[position] == '%')
+                        {
                             break;
+                        }
+                    }
 
                     if (position == stringToUnescape.Length)
+                    {
                         return stringToUnescape;
+                    }
 
                     UnescapeMode unescapeMode = UnescapeMode.Unescape | UnescapeMode.UnescapeAll;
                     position = 0;
-                    char[] dest = new char[stringToUnescape.Length];
-                    dest = UriHelper.UnescapeString(stringToUnescape, 0, stringToUnescape.Length, dest, ref position,
-                        c_DummyChar, c_DummyChar, c_DummyChar, unescapeMode, null, false);
-                    return new string(dest, 0, position);
+                    PooledCharArray pooledArray = new PooledCharArray(stringToUnescape.Length);
+                    UriHelper.UnescapeString(stringToUnescape, 0, stringToUnescape.Length, pooledArray, ref position,
```

Comment:
This looks broken.  PooledCharArray is a struct, so it'll be copied by value into UnescapeString.  If UnescapeString then calls GrowAndCopy, it'll be replacing the array in the copy that was passed in, but not this original.  Which means a) the resulting string that gets returned will be incorrect, and b) we'll be releasing the same array back into the ArrayPool twice.

This suggests to me we have a test hole in the tests that exercise this stuff.  I suggest before pursuing this further, we ensure those holes are filled, at which point I expect this code will cause failures.

In any event, using a struct like this requires a lot of care, and typically the way it's handled is passing the struct around by ref rather than by value, though even then it requires a lot of scrutiny to ensure it's never being copied inappropriately.